### PR TITLE
Add formatted logging for hyperdrive crash reports.

### PR DESCRIPTION
**Pull request by sentilesdal**
This PR adds some crash logging capabilities for hyperdrive.  In addition, `setup_logging` was refactored and tested, as was `format_float_as_string`.  In order to add hyperdrive crash logging, you would just add the line below any setup_logging call:
```python
setup_logging()
# add this line
setup_hyperdrive_crash_report_logging()
```

then, when logging errors, you'd also add a call to `log_hyperdrive_crash_report`:
```python
    except TransactionError as err:
        formatted_amount = format_utils.format_float_as_string(amount)
        pool_info = hyperdrive_contract.getPoolInfo()
        pool_config = hyperdrive_contract.getPoolConfig()
        logging.error(
            "Failed to execute %s: %s\n => Amount: %s\n => Agent: %s\n => Pool: %s",
            trade_type,
            err,
            formatted_amount,
            agent,
            dict(pool_info),
        )
        # add this!
        log_hyperdrive_crash_report(
            amount=amount,
            trade_type=trade_type,
            error=err,
            agent_address=agent.address,
            pool_info=pool_info,
            pool_config=pool_config,
        )
        raise err
```

this will get you logs to `.logging/hyperdrive_test_crash_report.log` that look like this:
```
23-07-06 12:21:01: CRITICAL: logs.log_hyperdrive_crash_report:
Failed to execute CLOSE_LONG: Message
 Amount: 1.23
 Agent: 0x0000000000000000000000000000000000000000
 PoolInfo: {
    "blockNumber": 1234,
    "timestamp": 12345678,
    "shareReserves": null,
    "bondReserves": null,
    "lpTotalSupply": null,
    "sharePrice": null,
    "longsOutstanding": null,
    "longAverageMaturityTime": null,
    "shortsOutstanding": null,
    "shortAverageMaturityTime": null,
    "shortBaseVolume": null,
    "withdrawalSharesReadyToWithdraw": null,
    "withdrawalSharesProceeds": null
}
 PoolConfig: {
    "contractAddress": "0x0000000000000000000000000000000000000000",
    "baseToken": null,
    "initializeSharePrice": null,
    "positionDuration": null,
    "checkpointDuration": null,
    "timeStretch": null,
    "governance": null,
    "feeCollector": null,
    "curveFee": null,
    "flatFee": null,
    "governanceFee": null,
    "oracleSize": null,
    "updateGap": null,
    "invTimeStretch": null,
    "termLength": null
}
```

### CODE REVIEW

File: elfpy/utils/outputs.py [hunk 1]
```diff
@@ -412,57 +412,6 @@ def annotate(axis_handle, text, major_offset, minor_offset, val):
...
-        for handler in logging.getLogger().handlers:
-            if hasattr(handler, "baseFilename") and not isinstance(handler, logging.StreamHandler):
-                # access baseFilename in a type safe way
-                handler_file_name = getattr(handler, "baseFilename", None)
-                if handler_file_name is not None and os.path.exists(handler_file_name):
-                    os.remove(handler_file_name)
-            handler.close()
-
-
 class ExtendedJSONEncoder(json.JSONEncoder):
```
Comment (dpaiton) [inline elfpy/utils/outputs.py:409]: I think it makes sense to move the JSONEncoder to logging as well.

Comment (dpaiton) [reply elfpy/utils/outputs.py:409]: And `str_with_precision`

File: elfpy/utils/logs.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+            Path and name of the log file.
+        max_bytes : (int, optional)
+            Maximum size of the log file in bytes. Defaults to elfpy.DEFAULT_LOG_MAXBYTES.
+        log_level : (int, optional)
+            Log level to track. Defaults to elfpy.DEFAULT_LOG_LEVEL.
+        delete_previous_logs : (bool, optional)
+            Whether to delete previous log file if it exists. Defaults to False.
+        log_file_and_stdout : (bool, optional)
+            Whether to log to both file and standard output. Defaults to False.
+        log_formatter (logging.Formatter, optional):
```
Comment (dpaiton) [inline elfpy/utils/logs.py:47]: colon in the wrong spot

**Review by slundqui — APPROVED** [review]
Looks great!

**Pull request merged**